lib/Builtins.cpp

```diff
@@ -14,171 +14,335 @@
 
 #include "Builtins.h"
 
+#include <stdlib.h>
+#include <unordered_map>
+
 using namespace llvm;
+using namespace clspv;
+
+namespace {
+////////////////////////////////////////////////////////////////////////////////
+////  Convert Builtin function name to a Type enum
+////////////////////////////////////////////////////////////////////////////////
+Builtins::BuiltinType LookupBuiltinType(const std::string &name) {
+
+// Build static map of builtin function names
+#include "BuiltinsMap.inc"
+
+  auto ii = s_func_map.find(name.c_str());
+  if (ii != s_func_map.end()) {
+    return (*ii).second;
+  }
+  return Builtins::kBuiltinNone;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+////  Mangled name parsing utilities
+////////////////////////////////////////////////////////////////////////////////
+
+// capture real name of function and struct types
+std::string GetUnmangledName(const std::string &str, size_t *pos) {
+  char *end;
+  int name_len = strtol(&str[*pos], &end, 10);
+  if (!name_len) {
+    return "";
+  }
+  size_t name_pos = end - str.data();
+
+  *pos = name_pos + name_len;
+  return str.substr(name_pos, name_len);
+}
+
+// capture parameter type and qualifiers based on Itanium ABI with OpenCL types
+// - return new parsing position pos, or zero for error
+size_t GetParameterType(const std::string &mangled_name,
+                        clspv::Builtins::ParamTypeInfo *type_info, size_t pos) {
+  // Parse a parameter type encoding
+  char type_code = mangled_name[pos++];
+
+  int blen = 1;
+  switch (type_code) {
+    // qualifiers
+  case 'P': // Pointer
+  case 'R': // Reference
+    return GetParameterType(mangled_name, type_info, pos);
+  case 'k': // ??? not part of cxxabi
+  case 'K': // const
+  case 'V': // volatile
+    return GetParameterType(mangled_name, type_info, pos);
+  case 'U': { // Address space
+    std::string address_space = GetUnmangledName(mangled_name, &pos);
+    return GetParameterType(mangled_name, type_info, pos);
+  }
+    // OCL types
+  case 'D':
+    type_code = mangled_name[pos++];
+    if (type_code == 'v') { // OCL vector
+      char *end;
+      int numElems = strtol(&mangled_name[pos], &end, 10);
+      if (!numElems) {
+        return 0;
+      }
+      type_info->vector_size = numElems;
+      pos = end - mangled_name.data();
+
+      if (mangled_name[pos++] != '_') {
+        return 0;
+      }
+      return GetParameterType(mangled_name, type_info, pos);
+    } else if (type_code == 'h') { // OCL half
+      type_info->type_id = Type::FloatTyID;
+      type_info->is_signed = true;
+      type_info->byte_len = 2;
+    } else {
+#ifdef DEBUG
+      printf("Func: %s\n", mangled_name.c_str());
+      llvm_unreachable("failed to demangle name");
+#endif
+      return 0;
+    }
+    break;
+
+    // element types
+  case 'l':
```

Comment:
I'd prefer no fallthrough cases. Might be clearer to separate length from type and signedness. Also please use `byte_len` instead `blen`

Reply:
Yes, I agree it is more complicated but slightly more efficient. For the sake of maintainability I will add another switch statement specifically for byte_len.